src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -54,17 +54,35 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
     workInProgress.pendingWorkPriority = NoWork;
   }
 
-  function updateClassComponent(current, workInProgress) {
+  function updateClassComponent(current : ?Fiber, workInProgress : Fiber) {
     var props = workInProgress.pendingProps;
-    if (!workInProgress.stateNode) {
+    var instance = workInProgress.stateNode;
+    if (!instance) {
       var ctor = workInProgress.type;
-      workInProgress.stateNode = new ctor(props);
+      workInProgress.stateNode = instance = new ctor(props);
+    } else if (typeof instance.shouldComponentUpdate === 'function') {
```

Comment:
nit: in ReactCompositeComponent we do `if (inst.shouldComponentUpdate)` which fails on non-functions whereas yours silently ignores them